blueprints/layered-profile-server.md

```diff
@@ -0,0 +1,296 @@
+
+- The number of Profiles quickly becomes a combinatorial explosion
+- It's nearly impossible for Ops engineers to figure out what Parameters are assigned to all Profiles of a given CDN, machine, or datacenter.
+- What about that one random Parameter on a Cloned profile that was changed a year ago? Is it still in place? Should it be? Did it need to be applied to all new Profiles cloned from this one?
+
+## Proposed Change
+
+Layered Profiles allow assigning multiple profiles to Servers. If multiple profiles have a parameter with the same name and config file, the parameter from the last profile in the ordering is applied.
+
+Layered Profiles is exactly as powerful as the existing profiles, it doesn't enable any new things. It makes profiles much easier to manage.
+
+With Layered Profiles, hundreds of profiles become a few dozen, each representing a logical group. For example, a server might have the profiles, in order:
+1. EDGE
+2. AMIGA_123
+3. CDN_FOO
+4. DATACENTER_ABC
+5. ATS_714
+6. Custom_Hack_42
+
+### Traffic Portal Impact
+
+1. A UI to view all parameters currently applied to a server, as well as the profile each parameter came from. A new page, linked from Server pages. For eg:
+
+| Name                                             |Config File               | Value                                | Profile   |
+|--------------------------------------------------|--------------------------|--------------------------------------|-----------|
+| location                                         | url_sig_myds.config      | /opt/trafficserver/etc/trafficserver | EDGE      |
+| Drive_Prefix                                     | storage.config           | /dev/sd                              | AMIGA_123 |
+| error_url                                        | url_sig_myotherds.config | 403                                  | EDGE      |
+| CONFIG proxy.config.exec_thread.autoconfig.scale | records.config           | FLOAT 1.5                            | ATS_714   |
+
+2. A UI change to add, remove, and reorder (sortable list) profiles for Servers, on the existing Server pages.
+3. Filtering based on Profiles will also need to be updated to take into account the plurality of Profiles.
+
+### Traffic Ops Impact
+
+- Add backend logic to the below-mentioned existing API endpoints, in order to show/create/update/delete a sorted list of profiles for server.
+
+- `/servers/  GET, POST`
+- `/servers/{id} PUT, DELETE`
+- `/servers/details`
+- `/deliveryservices/{{ID}}/servers`
+- `/deliveryservices/{{ID}}/servers/eligible`
+
+#### REST API Impact
+
+- No new endpoints are required
+
+**Existing Endpoints**
+
+- Modify JSON request and response for existing servers endpoints.
+- **_Note_**: All fields in the response structure for API 4.0/ 5.0 except `profileId`, `profileDesc` and `profile`, remain same. 
+  - In API 4.0/5.0 `profile` changes to `profiles` and in API 5.0 `profileId` and `profileDesc` are no longer part of response structure
```

Comment:
TO API 4.0 is currently declared unstable, so we are free to make the breaking changes to TO API 4.0 without having to make further changes in TO API 5.0 (if there even ends up being one). Because of that, what this blueprint describes as API 5.0 vs API 4.0 should all just be done in 4.0. There is no need to even mention TO API 5.0 at this point.

That said, t3c may require backwards-compatibility support similar to what was done for the `/jobs` breaking changes in API 4.0. Basically, it would need to handle both the _original_ 4.0 server struct (prior to implementing this blueprint) and the _new_ 4.0 server struct. It should attempt to unmarshal into the _new_ 4.0 struct first, then fall back to unmarshalling into the _original_ 4.0 struct if it determines that TO is not yet providing the _new_ 4.0 struct.

Reply:
I hought the agreed upon direction was to make the API output a list of length one for a full major version, then move to an arbitrary length list only once all API versions that use `profile`/`profileName`/`profileId` have been _fully removed_? Because the question was raised of how to handle backwards comparability for those old APIs. Since APIv4 is unstable, API version 3 is not yet deprecated, so it can't be removed until a full major ATC version passes where APIv4 is stable and APIv3 is deprecated.

Reply:
We definitely can and should just get rid of `profileId` and `profileDesc` immediately in APIv4, though

Reply:
Yes, we should permit a server to only have a single profile until API v3 has been deprecated and removed. In theory, if we stabilized 4.0 (with these changes) and deprecated 3.x in ATC 6.1, we could remove 3.x and allow servers to have more than 1 profile in the next release (e.g. 7.0). That wouldn't require a new 5.0 API version.

Reply:
The type changing from `[string]` to `[string, ...string[]]` isn't breaking at the parse level, but it is very different conceptually. Wouldn't people's scripts that look up a server's Profile be broken at that point? That is they'd silently give back incorrect results since they'd only expect to have to look at a single Profile? I guess on a certain level as long as this is communicated properly that'd be their fault, though.

Reply:
Once the data type is an array, any client/script that uses the array must be able to account for the fact that there can be more than 1 profile at that point. We might not _allow_ more than 1 profile when they start using it, but their code must assume layered profiles nonetheless. They can't/shouldn't rely on the fact that we'll only be allowing 1 profile until some point in the future.

Reply:
that actually makes more sense, now that I'm thinking about it. Done my way, we'd just be putting off the same problem until the APIv4/v5 transition point.

Reply:
This is really confusing me. I thought the plan was to implement the solution across two API end points to maintain backward compatibility. Now, I am reading that API 4.0 is unstable and we can totally implement this in one API call with a breaking change. The whole reason I have multiple comments on this blueprint was because I mentioned that the API was going to be a breaking change and may not maintain backward compatibility.

Reply:
I apologize. This is related to a mailing list proposal that I proposed back in late August after this blueprint was first submitted. It was [started here](https://lists.apache.org/thread/cth1xmmp5w85mkd1zoosvk55d1f622lv) and [continued here](https://lists.apache.org/thread/34v2v819yx658nls0gm5zdpslk22ozvn). Long story short: the TO API 4.0 is still considered "unstable" now. In other words, we are free to break it until such a time that we agree to consider it "stable," after which we will not make any more breaking changes to v4.0.

Before the unstable v4.0 API proposal was agreed upon, this blueprint was valid in stating that the v5.0 API would contain the new breaking changes. However, those breaking changes can now just be made in v4.0 instead, saving us from having to add yet another major TO API version.